File: crates/sdk/src/context.rs
use crate::{
    cli::{Cli, Command},
    system::get_system,
};
use anyhow::{bail, Result};
use clap::Parser;
use sha256::digest;
use std::collections::HashMap;
use tonic::{transport::Server, Code::NotFound, Request, Response, Status};
use tracing::info;
use vorpal_schema::{
    agent::v0::agent_service_client::AgentServiceClient,
    artifact::v0::{
        artifact_service_client::ArtifactServiceClient,
        artifact_service_server::{ArtifactService, ArtifactServiceServer},
        Artifact, ArtifactRequest, ArtifactResponse, ArtifactSystem, ArtifactsRequest,
        ArtifactsResponse,
    },
};

#[derive(Clone)]
pub struct ConfigContextStore {
    artifact: HashMap<String, Artifact>,
    variable: HashMap<String, String>,
}

#[derive(Clone)]
pub struct ConfigContext {
    agent: String,
    artifact: String,
    port: u16,
    registry: String,
    store: ConfigContextStore,
    system: ArtifactSystem,
}

#[derive(Clone)]
pub struct ArtifactServer {
    pub store: ConfigContextStore,
}

impl ArtifactServer {
    pub fn new(store: ConfigContextStore) -> Self {
        Self { store }
    }
}

#[tonic::async_trait]
impl ArtifactService for ArtifactServer {
    async fn get_artifact(
        &self,
        request: Request<ArtifactRequest>,
    ) -> Result<Response<Artifact>, Status> {
        let request = request.into_inner();

        if request.digest.is_empty() {
            return Err(tonic::Status::invalid_argument("'digest' is required"));
        }

        let artifact = self.store.artifact.get(request.digest.as_str());

        if artifact.is_none() {
            return Err(tonic::Status::not_found("artifact not found"));
        }

        Ok(Response::new(artifact.unwrap().clone()))
    }

    async fn get_artifacts(
        &self,
        _: tonic::Request<ArtifactsRequest>,
    ) -> Result<tonic::Response<ArtifactsResponse>, tonic::Status> {
        let response = ArtifactsResponse {
            digests: self.store.artifact.keys().cloned().collect(),
        };

        Ok(Response::new(response))
    }

    async fn store_artifact(
        &self,
        _request: Request<Artifact>,
    ) -> Result<Response<ArtifactResponse>, Status> {
        Err(Status::unimplemented("not implemented yet"))
    }
}

pub async fn get_context() -> Result<ConfigContext> {
    let args = Cli::parse();

    match args.command {
        Command::Start {
            agent,
            artifact,
            port,
            registry,
            target,
            variable,
        } => Ok(ConfigContext::new(
            agent, artifact, port, registry, target, variable,
        )?),
    }
}

impl ConfigContext {
    pub fn new(
        agent: String,
        artifact: String,
        port: u16,
        registry: String,
        system: String,
        variable: Vec<String>,
    ) -> Result<Self> {
        let store = ConfigContextStore {
            artifact: HashMap::new(),
            variable: variable
                .iter()
                .map(|v| {
                    let mut parts = v.split('=');
                    let name = parts.next().unwrap_or_default();
                    let value = parts.next().unwrap_or_default();
                    (name.to_string(), value.to_string())
                })
                .collect(),
        };

        let system = get_system(&system)?;

        Ok(Self {
            agent,
            artifact,
            port,
            registry,
            store,
            system,
        })
    }

    pub async fn add_artifact(&mut self, artifact: &Artifact) -> Result<String> {
        let artifact_json =
            serde_json::to_vec(artifact).expect("failed to serialize artifact to JSON");

        let artifact_digest = digest(artifact_json);

        if self.store.artifact.contains_key(&artifact_digest) {
            return Ok(artifact_digest);
        }

        // TODO: make this run in parallel

        let mut client = AgentServiceClient::connect(self.agent.clone())
            .await
            .expect("failed to connect to agent service");

        let response = client
            .prepare_artifact(artifact.clone())
            .await
            .expect("failed to prepare artifact");

        let mut response = response.into_inner();
        let mut response_artifact = None;
        let mut response_artifact_digest = None;

        loop {
            match response.message().await {
                Ok(Some(message)) => {
                    if let Some(artifact_output) = message.artifact_output {
                        if self.port == 0 {
                            info!("{} |> {}", artifact.name, artifact_output);
                        } else {
                            println!("{} |> {}", artifact.name, artifact_output);
                        }
                    }

                    response_artifact = message.artifact;
                    response_artifact_digest = message.artifact_digest;
                }
                Ok(None) => break,
                Err(status) => {
                    if status.code() != NotFound {
                        bail!("{}", status.message());
                    }

                    break;
                }
            }
        }

        if response_artifact.is_none() {
            bail!("artifact not returned from agent service");
        }

        if response_artifact_digest.is_none() {
            bail!("artifact digest not returned from agent service");
        }

        let artifact = response_artifact.unwrap();
        let artifact_digest = response_artifact_digest.unwrap();

        if !self.store.artifact.contains_key(&artifact_digest) {
            self.store
                .artifact
                .insert(artifact_digest.clone(), artifact.clone());
        }

        Ok(artifact_digest)
    }

    pub async fn fetch_artifact(&mut self, digest: &str) -> Result<String> {
        if self.store.artifact.contains_key(digest) {
            return Ok(digest.to_string());
        }

        let mut client = ArtifactServiceClient::connect(self.registry.clone())
            .await
            .expect("failed to connect to artifact service");

        let request = ArtifactRequest {
            digest: digest.to_string(),
        };

        match client.get_artifact(request.clone()).await {
            Err(status) => {
                if status.code() != NotFound {
                    bail!("artifact service error: {:?}", status);
                }

                bail!("artifact not found: {digest}");
            }

            Ok(response) => {
                let artifact = response.into_inner();

                self.store
                    .artifact
                    .insert(digest.to_string(), artifact.clone());

                for step in artifact.steps.iter() {
                    for artifact_digest in step.artifacts.iter() {
                        Box::pin(self.fetch_artifact(artifact_digest)).await?;
                    }
                }

                Ok(digest.to_string())
            }
        }
    }

    pub fn get_artifact_store(&self) -> HashMap<String, Artifact> {
        self.store.artifact.clone()
    }

    pub fn get_artifact(&self, digest: &str) -> Option<Artifact> {
        self.store.artifact.get(digest).cloned()
    }

    pub fn get_artifact_name(&self) -> &str {
        self.artifact.as_str()
    }

    pub fn get_target(&self) -> ArtifactSystem {
        self.system
    }

    pub fn get_variable(&self, name: &str) -> Option<String> {
        self.store.variable.get(name).cloned()
    }

    pub async fn run(&self) -> Result<()> {
        let service = ArtifactServiceServer::new(ArtifactServer::new(self.store.clone()));

        let service_addr_str = format!("[::]:{}", self.port);
        let service_addr = service_addr_str.parse().expect("failed to parse address");

        println!("artifact service: {}", service_addr_str);

        Server::builder()
            .add_service(service)
            .serve(service_addr)
            .await
            .map_err(|e| anyhow::anyhow!("failed to serve: {}", e))
    }
}

File: sdk/go/internal/config/context.go
package config

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"strings"

	agentApi "github.com/ALT-F4-LLC/vorpal/sdk/go/api/v0/agent"
	artifactApi "github.com/ALT-F4-LLC/vorpal/sdk/go/api/v0/artifact"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type ArtifactSource struct {
	Excludes []string
	Hash     *string
	Includes []string
	Path     string
}

type ArtifactSourceKind string

const (
	ArtifactSourceKind_GIT     ArtifactSourceKind = "GIT"
	ArtifactSourceKind_HTTP    ArtifactSourceKind = "HTTP"
	ArtifactSourceKind_LOCAL   ArtifactSourceKind = "LOCAL"
	ArtifactSourceKind_UNKNOWN ArtifactSourceKind = "UNKNOWN"
)

type ConfigContextStore struct {
	artifact map[string]*artifactApi.Artifact
	variable map[string]string
}

type ConfigContext struct {
	agent    string
	artifact string
	port     int
	registry string
	store    ConfigContextStore
	system   artifactApi.ArtifactSystem
}

type ArtifactServer struct {
	artifactApi.UnimplementedArtifactServiceServer

	store ConfigContextStore
}

func NewArtifactServer(store ConfigContextStore) *ArtifactServer {
	return &ArtifactServer{
		store: store,
	}
}

func (s *ArtifactServer) GetArtifact(ctx context.Context, request *artifactApi.ArtifactRequest) (*artifactApi.Artifact, error) {
	if request.Digest == "" {
		return nil, fmt.Errorf("'digest' is required")
	}

	response := s.store.artifact[request.Digest]
	if response == nil {
		return nil, fmt.Errorf("artifact not found")
	}

	return response, nil
}

func (s *ArtifactServer) GetArtifacts(ctx context.Context, request *artifactApi.ArtifactsRequest) (*artifactApi.ArtifactsResponse, error) {
	digests := make([]string, 0)

	for digest := range s.store.artifact {
		digests = append(digests, digest)
	}

	response := &artifactApi.ArtifactsResponse{
		Digests: digests,
	}

	return response, nil
}

func (s *ArtifactServer) StoreArtifact(ctx context.Context, request *artifactApi.Artifact) (*artifactApi.ArtifactResponse, error) {
	return nil, fmt.Errorf("not implemented")
}

func GetContext() *ConfigContext {
	cmd, err := newCommand()
	if err != nil {
		log.Fatal(err)
	}

	store := ConfigContextStore{
		artifact: make(map[string]*artifactApi.Artifact),
		variable: cmd.Variable,
	}

	return &ConfigContext{
		agent:    cmd.Agent,
		artifact: cmd.Artifact,
		port:     cmd.Port,
		registry: cmd.Registry,
		store:    store,
		system:   cmd.Target,
	}
}

func (c *ConfigContext) AddArtifact(artifact *artifactApi.Artifact) (*string, error) {
	// 1. Setup systems

	artifactJson, err := json.Marshal(artifact)
	if err != nil {
		return nil, err
	}

	artifactDigest := fmt.Sprintf("%x", sha256.Sum256(artifactJson))

	if _, ok := c.store.artifact[artifactDigest]; ok {
		return &artifactDigest, nil
	}

	// TODO: make this run in parallel

	agent := strings.ReplaceAll(c.agent, "http://", "")

	clientConn, err := grpc.NewClient(agent, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}

	defer clientConn.Close()

	client := agentApi.NewAgentServiceClient(clientConn)

	clientResponse, err := client.PrepareArtifact(context.Background(), artifact)
	if err != nil {
		return nil, fmt.Errorf("error preparing artifact: %v", err)
	}

	for {
		response, err := clientResponse.Recv()
		if err == io.EOF {
			break
		}

		if err != nil {
			return nil, fmt.Errorf("error receiving response: %v", err)
		}

		if response.ArtifactOutput != nil {
			output := fmt.Sprintf("%s |> %s", artifact.Name, *response.ArtifactOutput)
			println(output)
		}

		if response.Artifact != nil {
			artifact = response.Artifact
		}

		if response.ArtifactDigest != nil {
			artifactDigest = *response.ArtifactDigest
		}
	}

	if _, ok := c.store.artifact[artifactDigest]; !ok {
		c.store.artifact[artifactDigest] = artifact
	}

	return &artifactDigest, nil
}

func fetchArtifacts(client artifactApi.ArtifactServiceClient, digest string, store map[string]*artifactApi.Artifact) error {
	if _, ok := store[digest]; ok {
		return nil
	}

	clientResponse, err := client.GetArtifact(context.Background(), &artifactApi.ArtifactRequest{Digest: digest})
	if err != nil {
		return fmt.Errorf("error fetching artifact: %v", err)
	}

	if _, ok := store[digest]; !ok {
		store[digest] = clientResponse
	}

	for _, step := range clientResponse.Steps {
		if step != nil {
			for _, digest := range step.Artifacts {
				fetchArtifacts(client, digest, store)
			}
		}
	}

	return nil
}

func (c *ConfigContext) FetchArtifact(digest string) (*string, error) {
	if _, ok := c.store.artifact[digest]; ok {
		return &digest, nil
	}

	registry := strings.ReplaceAll(c.registry, "http://", "")

	clientConn, err := grpc.NewClient(registry, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}

	defer clientConn.Close()

	client := artifactApi.NewArtifactServiceClient(clientConn)

	err = fetchArtifacts(client, digest, c.store.artifact)
	if err != nil {
		return nil, fmt.Errorf("error fetching '%s': %v", digest, err)
	}

	return &digest, nil
}

func (c *ConfigContext) GetArtifact(digest string) *artifactApi.Artifact {
	return c.store.artifact[digest]
}

func (c *ConfigContext) GetArtifactName() string {
	return c.artifact
}

func (c *ConfigContext) GetTarget() artifactApi.ArtifactSystem {
	return c.system
}

func (c *ConfigContext) GetVariable(name string) *string {
	if _, ok := c.store.variable[name]; !ok {
		return nil
	}

	value := c.store.variable[name]

	return &value
}

func (c *ConfigContext) Run() error {
	var grpcServerOpts []grpc.ServerOption

	grpcServer := grpc.NewServer(grpcServerOpts...)

	artifactApi.RegisterArtifactServiceServer(grpcServer, NewArtifactServer(c.store))

	listenerAddr := fmt.Sprintf("[::]:%d", c.port)

	listener, err := net.Listen("tcp", listenerAddr)
	if err != nil {
		log.Fatalf("failed to listen: %v", err)
	}

	log.Printf("artifact service: %s", listenerAddr)

	err = grpcServer.Serve(listener)
	if err != nil {
		return err
	}

	return nil
}
